File: SIMAR/Lexer.java
package SIMAR;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Lexer {
    private ArrayList<String> li_exprReg;
    private ArrayList<String> li_nombre;
    private ArrayList<Short> li_tipo;
    private ArrayList<Token> li_Tokens;
    private ArrayList<String> li_PalabraReservada;
    private int numero_Linea = 1;
    private String texto;
    
    private static final String[] ST_EXPR = {
        "\\n", "\\s+", "[a-zA-Z][a-zA-Z0-9_]*", "[0-9]+\\.[0-9]*", "[0-9]+",
        "[.]|[+]|[-]|[*]|[;]|[==]|[<]|[>]|[=]|[!=]|[!]|[#]|[\"]|[\\/]",
        "[(]", "[)]", "[{]", "[}]"
    };
    
    private static final String[] ST_NOMBRE = {
        "Salto de linea", "ws", "identificador", "flotante", "entero",
        "operador", "op_parentAbre", "op_parentCierra", "op_llaveAbre", "op_llaveCierra"
    };
    
    public Lexer() {
        li_exprReg = new ArrayList<>();
        li_nombre = new ArrayList<>();
        li_tipo = new ArrayList<>();
        li_Tokens = new ArrayList<>();
        li_PalabraReservada = new ArrayList<>();
    }
    
    public void cargarDefault() {
        for (int i = 0; i < ST_NOMBRE.length; i++) {
            li_exprReg.add(ST_EXPR[i]);
            li_nombre.add(ST_NOMBRE[i]);
            li_tipo.add((short)i);
        }
    }
    
    public void cargarArchivo(String archivo) {
        this.texto = archivo;
    }
    
    public boolean analizarArchivo() {
        preparar();
        int elements = li_exprReg.size();
        int caracteres = texto.length();
        boolean isOk = true;
        
        while (caracteres > 0 && isOk) {
            boolean matched = false;
            for (int i = 0; i < elements; i++) {
                Pattern pattern = Pattern.compile("^" + li_exprReg.get(i));
                Matcher matcher = pattern.matcher(texto);
                
                if (matcher.find()) {
                    matched = true;
                    short tipo = (short)i;
                    String stg = matcher.group();
                    texto = texto.substring(stg.length());
                    caracteres = texto.length();
                    String nombre = li_nombre.get(i);
                    
                    if (tipo == 0) {
                        numero_Linea++;
                        break;
                    }
                    
                    if (!stg.trim().isEmpty()) {
                        li_Tokens.add(new Token(nombre, stg, tipo));
                        break;
                    }
                }
            }
            if (!matched) isOk = false;
        }
        return isOk;
    }
    
    public String listarTokens() {
        StringBuilder listaTokens = new StringBuilder();
        for (Token token : li_Tokens) {
            listaTokens.append(token.tipo).append("\t")
                       .append(token.nombre).append("\t")
                       .append(token.tk).append("\n");
        }
        return listaTokens.toString();
    }
    
    public void agregarExprReg(String regExpr, String nombre) {
        li_exprReg.add(regExpr);
        li_nombre.add(nombre);
        short i = (short)li_nombre.indexOf(nombre);
        li_tipo.add(i);
    }
    
    public boolean agregarPalabraReservada(String st) {
        Pattern pattern = Pattern.compile(ST_EXPR[2]);
        Matcher matcher = pattern.matcher(st);
        boolean b = matcher.matches();
        if (b) {
            b = esPalabraReservada(st);
            if (!b) {
                li_PalabraReservada.add(st);
            }
        }
        return true;
    }
    
    public int Linea() {
        return numero_Linea;
    }
    
    public boolean esPalabraReservada(String st) {
        return li_PalabraReservada.contains(st);
    }
    
    private void preparar() {
        numero_Linea = 0;
        li_Tokens.clear();
    }
    
    public List<Token> getList() {
        return li_Tokens;
    }
}

File: SIMAR/Python/Lexer.py
# Lexer.py

import re

class Lexer:
    def __init__(self):
        self.li_exprReg = []
        self.li_nombre = []
        self.li_tipo = []
        self.li_Tokens = []
        self.li_PalabraReservada = []
        self.numero_Linea = 1
        self.texto = ""

        self.ST_EXPR = [
            r"^\n",
            r"^\s+",
            r"^[a-zA-Z]+_?[a-zA-Z0-9]*",
            r"^\d+\.\d*",
            r"^\d+",
            r"^[.]|^[+]|^-|^[*]|^[;]|^[=][=]|^[<]|^[>]|^[=]|^[!][=]|^[!]|^[#]|^[\"]|^[\\/]",
            r"^[(]",
            r"^[)]",
            r"^[{]",
            r"^[}]"
        ]
        self.ST_NOMBRE = [
            "Salto de linea",
            "ws",
            "identificador",
            "flotante",
            "entero",
            "operador",
            "op_parentAbre",
            "op_parentCierra",
            "op_llaveAbre",
            "op_llaveCierra"
        ]

    def cargarDefault(self):
        for i, expr in enumerate(self.ST_EXPR):
            self.li_exprReg.append(expr)
            self.li_nombre.append(self.ST_NOMBRE[i])
            self.li_tipo.append(i)

    def cargarArchivo(self, archivo):
        self.texto = archivo

    def analizarArchivo(self):
        self.preparar()
        elementos = len(self.li_exprReg)
        caracteres = len(self.texto)
        isOk = True
        texto_temp = self.texto

        while caracteres > 0 and isOk:
            encontrado = False
            for i in range(elementos):
                patron = self.li_exprReg[i]
                tipo = i
                match = re.match(patron, texto_temp)
                if match:
                    encontrado = True
                    indice = match.end()
                    stg = texto_temp[:indice]
                    texto_temp = texto_temp[indice:]
                    caracteres = len(texto_temp)
                    nombre = self.li_nombre[i]

                    if tipo == 0:
                        self.numero_Linea += 1
                        break
                    elif not stg.isspace() or stg == ' ':
                        if not (len(stg) == 1 and stg.isspace()):
                            self.li_Tokens.append(Token(nombre, stg, tipo))
                        break
                    else:
                        break
            if not encontrado:
                isOk = False

        self.texto = texto_temp
        return isOk

    def listarTokens(self):
        lista = ""
        for token in self.li_Tokens:
            lista += f"{token.tipo}\t{token.nombre}\t{token.tk}\n"
        return lista

    def agregarExprReg(self, regExpr, nombre):
        self.li_exprReg.append(regExpr)
        self.li_nombre.append(nombre)
        tipo = len(self.li_nombre) - 1
        self.li_tipo.append(tipo)

    def agregarPalabraReservada(self, st):
        match = re.match(self.ST_EXPR[2], st)
        if match and match.group() == st:
            if not self.esPalabraReservada(st):
                self.li_PalabraReservada.append(st)
                return True
        return False

    def Linea(self):
        return self.numero_Linea

    def esPalabraReservada(self, st):
        return st in self.li_PalabraReservada

    def preparar(self):
        self.numero_Linea = 0
        self.li_Tokens = []

    def getList(self):
        return self.li_Tokens

from Token import Token